### ISSUE TITLE
In-app messages vs Original in-app messages

### ISSUE THREAD

**Issue by istalev** [Appboy/appboy-android-sdk#61]
In terms of SDK usage and receiving in-app messages, is there a difference between in-app messages and (the deprecated) original in-app messages? 

The original in-app messages work fine, but the regular in-app messages are not displaying at all. I tried the latest 1.15.3 SDK version as well.

Also just curious why original in-app messages are being deprecated since it seems they have more flexibility in terms of delivery options. Thanks!

**Comment by Bucimis** [comment]
Hi @istalev,

Thanks for getting in touch!

Original in-app messages display automatically once they are received, typically right after session start.

On the other hand, regular in-app messages can be triggered for display on session start, by a custom event or purchase (including properties), or by a user clicking on a push notification.  As a result, the display of in-app messages can be more finely controlled than with original in-app messages.

Note that in order to get a regular in-app message to display, you'll need to define a trigger event (such as session start) that causes the event to display. Then, to display the message, that trigger event will need to occur. In the case of session start, the user just needs to start a new session (open the app).

You should also be able to sanity check regular in-app message display using our test message feature (click the "eye" icon on the top right of the in-app message composer).

If you continue to experience problems with regular in-app messages, feel free to send over logcat or any other information to help us identify, repro, and fix the issue.

Thanks,
Jared

**Comment by istalev** [comment]
Hey @Bucimis,

I appreciate your quick response!

Using a campaign to display an original in-app message with the "Start Session In Any App" trigger works fine. But when I try to do the same with a regular in-app message, I never receive the message.

From looking at the 1.15.2 and 1.15.3 change-logs, I thought perhaps you guys were fixing a similar issue?

Also, the "Scheduled Delivery" delivery method seems like a useful feature, but regular in-app messages only support "Action-Based Delivery".

**Comment by Bucimis** [comment]
Hey @istalev,

With original in-app messages, schedule-based delivery means that in-app messages are delivered when the SDK calls `requestInAppMessageRefresh()` within the scheduled delivery window. Typical integrations call `requestInAppMessageRefresh()` at session start, so this can be considered a session start-like delivery in most cases.

Action-based delivery means that in-app messages are delivered when the SDK calls `requestInAppMessageRefresh()`, provided the SDK has already performed the required action. Devices need to send up the triggering action (like a custom event) to the server before the in-app message is sent to the SDK, meaning there is a lag in delivery.

In both of these cases, the SDK does not store in-app messages locally. It requests them from the server manually and receives them if the user is eligible to receive the campaign and the scheduling criteria is met.

With new in-app messages, all in-app messages are delivered to the device and stored locally as soon as possible. When a triggering event occurs (like a custom event), the in-app message is available locally and can be shown right away. Due to this, action based delivery is much more useful and effective.

However, the local storage paradigm means that certain aspects of schedule-based delivery no longer makes sense (like sending at optimal times, or sending recurringly). Almost all of this functionality can be made up for through re-eligibility and other mechanisms.

If you have a specific scheduling use-case that you'd like to meet, we'd be happy to help you there.

--

As for the in-app message not showing, do you happen to have logs of the issue? (note - You can set Appboy logs to verbose by setting `com.appboy.support.AppboyLogger.LogLevel = Log.VERBOSE;`)

The bugfix in 1.15.3 is fairly niche and is likely unrelated, but logs will help illuminate whats going on.

Thanks,
Jared

**Comment by istalev** [comment]
Hey @Bucimis,

Thanks for all the detailed information! 

I was wrong; the new in-app message does show up. I had a few test campaigns running, testing both the original and new in-app messages. I was misled due to the metrics being reported on the campaign details page. For example, the new in-app message campaign shows "N/A" for the number of times the message was sent, whereas the original in-app message campaigns correctly display how many times the campaign was sent. Same for the Number of Unique Recipients: the original displays 1 (because I only targeted myself), where as the new in-app message shows 0 (even though I received it multiple times). The message has been displaying over the course of several days, so it seems like it might not be an issue with delay. 

So in other words, the new in-app message does display, but the metrics for it do not seem to be updated. 

Thanks again,

Ivan

**Comment by Bucimis** [comment]
Hey @istalev,

I see, thanks for the update.  A few questions: 

1) Are there any other metrics appearing for that campaign, such as impressions/button clicks/body clicks/etc? 
2) Is it possible to reproduce the lack of analytics on a separate (new) campaign?
3) More of a sanity check, but is the device connected to the internet?  If it received the messages earlier but has been offline since then, the analytics may not have reached our servers.  Messages can display in an offline state because they are displayed locally.

If you have a logcat of the moment the in-app message displays and is clicked/dismissed, that would be helpful as well.

Thanks,
Jared

**Comment by istalev** [comment]
Hey @Bucimis,

1) There are no metrics at all; everything seems to be at zero.
2) I created a new campaign with (the new) in-app message, and I still see no metrics. The campaign is targeted towards both Android and iOS. When it was viewed on iOS, the metrics appeared immediately :/
3) For sanity checks, I tested on several Android devices that always have an internet connection.

Scanning through logcat output, I don't see any output coming from Appboy before/after the in-app message was displayed, even after setting it to verbose. 

Again, the metrics for "original" in-app messages appear immediately :/

Thanks!

Ivan

**Comment by Bucimis** [comment]
Hi @istalev,

Sorry to hear you're still experiencing the issue. Could you possibly send us the verbose Appboy logcat output for a session with an in-app message display (from session start -> in-app message display -> close app/session close)?  

Note: Appboy logcat tags always start with "Appboy" - if you set output to verbose, there should be a high amount of logs coming through.

That will help us debug more fully what is going on on the client.  If the campaign data is correct and the client is sending up events properly, we'll need to take a look at the campaign on the server side.

Thanks,
Jared

**Comment by istalev** [comment]
Hey @Bucimis,

Here is the logcat:

D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the boolean configuration value with key com_appboy_disable_uil_image_cache. Using default value 'false'.
I/Appboy v1.15.0 .com.appboy.Appboy: Automatic GCM registration not enabled in appboy.xml. Appboy will not register for GCM.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the boolean configuration value with key com_appboy_push_adm_messaging_registration_enabled. Using default value 'false'.
I/Appboy v1.15.0 .com.appboy.Appboy: Automatic ADM registration not enabled in appboy.xml. Appboy will not register for ADM.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the string array configuration value with key com_appboy_locale_api_key_map. Using default value 'null'.
I/art: Rejecting re-init on previously-failed class java.lang.Class<com.appboy.services.AppboyWearableListenerService>
I/Appboy v1.15.0 .bo.app.cf: AppboyWearableListenerService not found on path. Service not available.
I/Appboy v1.15.0 .bo.app.cf: Appboy Wearable service is not available. Declare <service android:name="com.appboy.services.AppboyWearableListenerService"/> in your appboy.xml to enable Appboy wearable service. See the Droidboy manifest for an example
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the integer configuration value with key com_appboy_trigger_action_minimum_time_interval_seconds. Using default value '30'.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the boolean configuration value with key com_appboy_disable_location_collection. Using default value 'false'.
I/Appboy v1.15.0 .bo.app.by: Location collection enabled via appboy.xml configuration.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the boolean configuration value with key com_appboy_enable_background_location_collection. Using default value 'false'.
I/Appboy v1.15.0 .bo.app.by: Background location collection disabled via appboy.xml configuration.
I/Appboy v1.15.0 .bo.app.by: Appboy location service is not available. Declare <service android:name="com.appboy.services.AppboyLocationService"/> in your AndroidManifest.xml to enable Appboy location service.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the integer configuration value with key com_appboy_location_update_time_interval. Using default value '-1'.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the integer configuration value with key com_appboy_location_update_distance. Using default value '-1'.
D/Appboy v1.15.0 .com.appboy.Appboy: Appboy loaded in 427 ms.
D/Appboy v1.15.0 .com.appboy.configuration.CachedConfigurationProvider: Unable to find the boolean configuration value with key com_appboy_server_target. Using default value 'PROD'.
D/Appboy v1.15.0 .bo.app.k: Making request to [https://dev.appboy.com/api/v2/data], with headers: [(X-Appboy-User-Identifier / 641701948973254634), (Content-Type / application/json), (X-Appboy-Device-Identifier / 856158e1-4c7d-41bc-b188-f2a7454c7cc9), (Accept-Encoding / gzip, deflate)] and JSON parameters: [{"extras":{"sessions":[{"guid":"50fe25b0-21b8-4462-bdbc-b37b58e5dea0","start_time":1.477335548523E9,"new":true},{"guid":"ecf1ab49-a60c-4652-8fe1-637c7e437547","start_time":1.477332257499E9,"end_time":1.477333518938E9}],"device":{},"user":{}},"environment":{"sdk_version":"1.15.0","now":1.477335548732E9,"version_code":241,"version_name":"4.7.0","package_name":"me.lyft.android.dev","config_time":"1444757900","no_acks":true},"only_respond_with":["triggers"]}]
D/Appboy v1.15.0 .bo.app.k: Making request to [https://dev.appboy.com/api/v2/data], with headers: [(X-Appboy-User-Identifier / 641701948973254634), (Content-Type / application/json), (X-Appboy-Device-Identifier / 856158e1-4c7d-41bc-b188-f2a7454c7cc9), (Accept-Encoding / gzip, deflate)] and JSON parameters: [{"extras":{},"environment":{"sdk_version":"1.15.0","now":1.477335548762E9,"version_code":241,"version_name":"4.7.0","package_name":"me.lyft.android.dev","no_acks":true},"only_respond_with":["inapp"]}]
D/Appboy v1.15.0 .bo.app.l: Request Executed in [476ms] [POST:https://dev.appboy.com/api/v2/data]
D/Appboy v1.15.0 .bo.app.l: Request Executed in [516ms] [POST:https://dev.appboy.com/api/v2/data]
D/Appboy v1.15.0 .bo.app.bc: com.appboy.events.InAppMessageEvent fired: com.appboy.events.InAppMessageEvent@fe5d999
D/Appboy v1.15.0 .bo.app.bc: Triggering com.appboy.events.InAppMessageEvent on 1 subscribers.
I/Appboy v1.15.0 .bo.app.by: Single location update received from Appboy location manager: me.lyft.android.dev.SINGLE_APPBOY_LOCATION_UPDATE

**Comment by Bucimis** [comment]
Hey @istalev 

Thanks for sending this over.  Three quick asks:
1) Could you update to v1.15.3? That's our latest version and also contains more verbose logging than 1.15.0.
2) Could you turn off the call to `requestInAppMessageRefresh()`?  It will make it clearer that only triggered in-app messages are displaying.
3) For the logcat, it appears to cut off right when the in-app message is triggered. Would it be possible to get the entire log from session start to session end?  So the entire period from when the app is opened to 10 seconds after the app is closed:

- start recording logcat
-open app/session
-trigger an in-app message
-click on a button/message body
-close app
-wait 15 seconds
-stop recording

Apologies for being too vague earlier.  If the text gets too long for a comment, feel free to link to a dropbox file, gist, or other external resource.

Separately from this, another thing to try would be to call `requestImmediateDataFlush()` after an in-app message is clicked and seeing if that causes the events to populate on the dashboard. That may be helpful in the case of a long session timeout.

Thanks,
Jared

**Comment by istalev** [comment]
Hey @Bucimis,

1) Bumped version to 1.15.3
2) Here is a [gist](https://gist.github.com/istalev/63f812db72f13a5ad7440f1fb21833e5) of logcat output with `requestInAppMessageRefresh()` call removed. Here is a [gist](https://gist.github.com/istalev/bd67834114841f16d5d1728128f3f4d6) with calling `requestImmediateDataFlush()` when the in-app message is clicked.
3) The gists (I believe) include the entire session. I verified that `Appboy.getInstance(activity).closeSession(activity)` is called when the app is closed, but there is no additional logging than what is in the gists.

Apologies for my lack of understanding in regards to how the appboy SDK works; I wasn't the one to set it up so I haven't invested much time to debug. I appreciate your help!

Ivan

**Comment by Bucimis** [comment]
Hey @istalev 

Thanks for sending these logs.  They raised a few questions among engineers on our team and as a result, I was wondering if you could send over a debug APK with IAMs integrated to our team email at `android@appboy.com` - then we can debug e2e a little more closely.

Thanks again,
Jared

**Comment by Bucimis** [comment]
Also, could you send the campaign information? (names of campaigns) @istalev

**Comment by istalev** [comment]
Hey @Bucimis,

I'm getting some pushback regarding sending over a debug APK. Is there any other way we can help debug?

We do have somewhat of a custom integration with the appboy SDK, so perhaps it could be an issue with that. In the meantime, we can keep using the deprecated "original in-app messages" since the metrics work fine.

Ivan

**Comment by briancaw** [comment]
Hi @istalev 

Can you send a link to the campaign you're seeing this behavior on to android@appboy.com?  Also - can you let us know about the custom handling you have?  Also, once you send that email, let's chat about possibly setting up time to chat/screenshare if you're free; we plan on deprecating old-style in-app messages in the not to distant future so it'd be great to get things working for you/make sure triggered IAMs satisfy all of your requirements.

Thanks,
Brian

**Comment by Bucimis** [comment]
Hey @istalev,

Just following up, were you able to integrate new style triggered in-app messages successfully? We would be more than happy to jump on a call to discuss if there are still open questions.

Cheers,
Jared

**Comment by Bucimis** [comment]
Hi @istalev,

I'm going to close this issue out for now.  Please let us know if this is still an issue or if you have further comments/concerns.  As mentioned before, we'd be happy to jump on a call anytime to make sure everything is resolved and that you've been able to standardize on triggered IAMs.

Thanks!
